packages/react-ui/src/app/features/builder/run-details/loop-iteration-input.tsx

```diff
@@ -44,103 +71,178 @@ const LoopIterationInput = ({ stepName }: { stepName: string }) => {
     ) {
       setLoopIndex(stepName, totalIterations - 1);
     }
-  }, [totalIterations, currentIndex]);
+  }, [totalIterations, currentIndex, setLoopIndex, stepName]);
 
-  function onChange(e: React.ChangeEvent<HTMLInputElement>) {
+  const iterationOptions = useMemo(() => {
+    return Array.from({ length: totalIterations }, (_, i) => i + 1);
+  }, [totalIterations]);
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
     const value = e.target.value ?? '1';
     const parsedValue = Math.max(
       1,
       Math.min(parseInt(value) ?? 1, totalIterations),
     );
     setLoopIndex(stepName, parsedValue - 1);
-  }
+  };
+
+  const onBlur = (e: React.FocusEvent<HTMLInputElement>) => {
+    if (e.target.value === '' && inputRef.current) {
+      inputRef.current.value = '1';
+      setLoopIndex(stepName, 0);
+    }
+  };
 
-  function removeFocus() {
-    setIsFocused(false);
+  const handleSelectIteration = (iteration: number) => {
     if (inputRef.current) {
-      inputRef.current.blur();
+      inputRef.current.value = String(iteration);
     }
-    if (!isNil(inputRef.current) && inputRef.current.value.length === 0) {
-      setLoopIndex(stepName, 0);
+    setLoopIndex(stepName, iteration - 1);
+    setIsDropdownOpen(false);
+  };
+
+  const onInputClick = () => {
+    if (!isStepSelected) {
+      selectStepByName(stepName);
     }
-  }
+    setIsDropdownOpen(true);
+  };
+
+  const onInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'ArrowRight' && currentIndex < totalIterations - 1) {
+      setLoopIndex(stepName, currentIndex + 1);
+    }
+    if (e.key === 'ArrowLeft' && currentIndex > 0) {
+      setLoopIndex(stepName, currentIndex - 1);
+    }
+  };
+
+  const onArrowClick = (newIndex: number) => {
+    setLoopIndex(stepName, newIndex);
+    if (inputRef.current) {
+      inputRef.current.value = String(newIndex + 1);
+    }
+  };
```

Comment:
It would be good to use `useCallback` for function declarations, to optimize between renders